File: eva-ioni/src/sensors/environmental.js
/**
 * EVA IONI 2.0 - Environmental Sensors Module
 * Sensori ambientali supplementari per monitoraggio completo
 */

const EventEmitter = require('events');

class EnvironmentalSensor extends EventEmitter {
  constructor() {
    super();
    this.sensors = {
      airQuality: {
        pm25: 0,
        pm10: 0,
        co2: 0,
        voc: 0
      },
      weather: {
        temperature: 22,
        humidity: 60,
        windSpeed: 0,
        rain: false,
        uvIndex: 0
      },
      acoustic: {
        level: 0,
        birdSongs: 0,
        noisePollution: 0
      },
      light: {
        par: 0,
        lux: 0
      }
    };
    this.history = [];
    this.updateInterval = null;
  }

  startMonitoring(interval = 5000) {
    this.updateInterval = setInterval(() => {
      this.readAllSensors();
    }, interval);
    console.log('🌍 Environmental sensors started');
  }

  stopMonitoring() {
    if (this.updateInterval) {
      clearInterval(this.updateInterval);
      this.updateInterval = null;
    }
    console.log('🌍 Environmental sensors stopped');
  }

  readAllSensors() {
    // Simula lettura sensori
    this.sensors.airQuality = {
      pm25: 5 + Math.random() * 10,
      pm10: 10 + Math.random() * 20,
      co2: 400 + Math.random() * 200,
      voc: 0.1 + Math.random() * 0.5
    };

    this.sensors.weather = {
      temperature: 18 + Math.random() * 14,
      humidity: 40 + Math.random() * 40,
      windSpeed: Math.random() * 10,
      rain: Math.random() > 0.8,
      uvIndex: Math.floor(Math.random() * 11)
    };

    this.sensors.acoustic = {
      level: 30 + Math.random() * 30,
      birdSongs: Math.floor(Math.random() * 20),
      noisePollution: Math.random() * 10
    };

    this.sensors.light = {
      par: 100 + Math.random() * 800,
      lux: 1000 + Math.random() * 8000
    };

    this.history.push({
      timestamp: new Date().toISOString(),
      ...this.sensors
    });

    this.emit('sensor-data', this.sensors);
    this.emit('air-quality', this.sensors.airQuality);
    this.emit('weather', this.sensors.weather);
    this.emit('acoustic', this.sensors.acoustic);
    this.emit('light', this.sensors.light);

    return this.sensors;
  }

  getCurrentData() {
    return {
      timestamp: new Date().toISOString(),
      ...this.sensors
    };
  }

  getHistory(limit = 100) {
    return this.history.slice(-limit);
  }

  getAirQuality() {
    const aq = this.sensors.airQuality;
    let quality = 'good';
    if (aq.pm25 > 35 || aq.pm10 > 50) quality = 'moderate';
    if (aq.pm25 > 55 || aq.pm10 > 150) quality = 'unhealthy';
    if (aq.pm25 > 150 || aq.pm10 > 250) quality = 'hazardous';
    return { ...aq, quality };
  }

  getWeatherForecast() {
    const weather = this.sensors.weather;
    let forecast = 'clear';
    if (weather.rain) forecast = 'rainy';
    if (weather.windSpeed > 30) forecast = 'windy';
    if (weather.temperature > 30) forecast = 'hot';
    if (weather.temperature < 5) forecast = 'cold';
    return { ...weather, forecast };
  }

  getBiodiversityScore() {
    const birds = this.sensors.acoustic.birdSongs;
    const noise = this.sensors.acoustic.noisePollution;
    const uv = this.sensors.weather.uvIndex;
    
    let score = 50;
    score += birds * 2;
    score -= noise * 2;
    score -= Math.max(0, uv - 5) * 3;
    
    return Math.max(0, Math.min(100, score));
  }

  async calibrateSensors() {
    // Simula calibrazione
    console.log('🔧 Calibrating environmental sensors...');
    await new Promise(resolve => setTimeout(resolve, 2000));
    console.log('✅ Sensors calibrated successfully');
    return { success: true, message: 'Calibration complete' };
  }
}

module.exports = new EnvironmentalSensor();

File: eva-ioni-python/src/sensors/environmental.py
"""
EVA IONI 2.0 - Environmental Sensors Module
Sensori ambientali supplementari per monitoraggio completo
"""

import random
import asyncio
from datetime import datetime
from typing import Dict, Any, List, Optional

class EnvironmentalSensor:
    def __init__(self):
        self.sensors = {
            'air_quality': {
                'pm25': 0.0,
                'pm10': 0.0,
                'co2': 0.0,
                'voc': 0.0
            },
            'weather': {
                'temperature': 22.0,
                'humidity': 60.0,
                'wind_speed': 0.0,
                'rain': False,
                'uv_index': 0
            },
            'acoustic': {
                'level': 0.0,
                'bird_songs': 0,
                'noise_pollution': 0.0
            },
            'light': {
                'par': 0.0,
                'lux': 0.0
            }
        }
        self.history = []
        self._callbacks = []

    def read_all_sensors(self) -> Dict[str, Any]:
        """Leggi tutti i sensori e aggiorna i dati"""
        # Simula lettura sensori
        self.sensors['air_quality'] = {
            'pm25': 5 + random.random() * 10,
            'pm10': 10 + random.random() * 20,
            'co2': 400 + random.random() * 200,
            'voc': 0.1 + random.random() * 0.5
        }

        self.sensors['weather'] = {
            'temperature': 18 + random.random() * 14,
            'humidity': 40 + random.random() * 40,
            'wind_speed': random.random() * 10,
            'rain': random.random() > 0.8,
            'uv_index': random.randint(0, 10)
        }

        self.sensors['acoustic'] = {
            'level': 30 + random.random() * 30,
            'bird_songs': random.randint(0, 20),
            'noise_pollution': random.random() * 10
        }

        self.sensors['light'] = {
            'par': 100 + random.random() * 800,
            'lux': 1000 + random.random() * 8000
        }

        timestamp = datetime.now().isoformat()
        data = {
            'timestamp': timestamp,
            **self.sensors
        }
        self.history.append(data)
        
        # Notify callbacks
        for callback in self._callbacks:
            callback(data)
        
        return data

    def get_current_data(self) -> Dict[str, Any]:
        """Ottieni i dati correnti"""
        return {
            'timestamp': datetime.now().isoformat(),
            **self.sensors
        }

    def get_history(self, limit: int = 100) -> List[Dict]:
        """Ottieni lo storico dei dati"""
        return self.history[-limit:]

    def get_air_quality(self) -> Dict[str, Any]:
        """Ottieni la qualità dell'aria"""
        aq = self.sensors['air_quality']
        quality = 'good'
        if aq['pm25'] > 35 or aq['pm10'] > 50:
            quality = 'moderate'
        if aq['pm25'] > 55 or aq['pm10'] > 150:
            quality = 'unhealthy'
        if aq['pm25'] > 150 or aq['pm10'] > 250:
            quality = 'hazardous'
        return {**aq, 'quality': quality}

    def get_weather_forecast(self) -> Dict[str, Any]:
        """Ottieni le previsioni meteo"""
        weather = self.sensors['weather']
        forecast = 'clear'
        if weather['rain']:
            forecast = 'rainy'
        if weather['wind_speed'] > 30:
            forecast = 'windy'
        if weather['temperature'] > 30:
            forecast = 'hot'
        if weather['temperature'] < 5:
            forecast = 'cold'
        return {**weather, 'forecast': forecast}

    def get_biodiversity_score(self) -> int:
        """Calcola il punteggio di biodiversità"""
        birds = self.sensors['acoustic']['bird_songs']
        noise = self.sensors['acoustic']['noise_pollution']
        uv = self.sensors['weather']['uv_index']
        
        score = 50
        score += birds * 2
        score -= noise * 2
        score -= max(0, uv - 5) * 3
        
        return max(0, min(100, score))

    def on_data(self, callback):
        """Registra un callback per i dati"""
        self._callbacks.append(callback)

    async def calibrate(self):
        """Calibra i sensori"""
        print("🔧 Calibrating environmental sensors...")
        await asyncio.sleep(2)
        print("✅ Sensors calibrated successfully")
        return {'success': True, 'message': 'Calibration complete'}

# Singleton
environmental = EnvironmentalSensor()
